### El problema de los misioneros y los caníbales en AIMA

In [2]:
# importamos las cosas que vamos a usar de aima
from search import *

class ProblemaMisioneros(Problem):
    ''' Clase problema (formalizacion de nuestro problema) siguiendo la
        estructura que aima espera que tengan los problemas.'''
    def __init__(self, initial, goal=None):
        '''Inicializacion de nuestro problema.'''
        Problem.__init__(self, initial, goal)
        # cada accion tiene un texto para identificar al operador y despues una tupla con la
        # cantidad de misioneros y canibales que se mueven en la canoa
        self._actions = [('1c', (0,1)), ('1m', (1, 0)), ('2c', (0, 2)), ('2m', (2, 0)), ('1m1c', (1, 1))]
        self.analizados  = 0

    def actions(self, s):
        '''Devuelve las acciones validas para un estado.'''
        # las acciones validas para un estado son aquellas que al aplicarse
        # nos dejan en otro estado valido
        return [a for a in self._actions if self._is_valid(self.result(s, a))]

    def _is_valid(self, s):
        '''Determina si un estado es valido o no.'''
        # un estado es valido si no hay mas canibales que misioneros en ninguna
        # orilla, y si las cantidades estan entre 0 y 3
        return (s[0] >= s[1] or s[0] == 0) and ((3 - s[0]) >= (3 - s[1]) or s[0] == 3) and (0 <= s[0] <= 3) and (0 <= s[1] <= 3)

    def goal_test(self, estado):
        self.analizados += 1
        return super().goal_test(estado)
    
    def result(self, s, a):
        '''Devuelve el estado resultante de aplicar una accion a un estado
           determinado.'''
        # el estado resultante tiene la canoa en el lado opuesto, y con las
        # cantidades de misioneros y canibales actualizadas segun la cantidad
        # que viajaron en la canoa
        if s[2] == 0:
            return (s[0] - a[1][0], s[1] - a[1][1], 1)
        else:
            return (s[0] + a[1][0], s[1] + a[1][1], 0)

    @staticmethod
    def evalAlgorithm(estado_inicial, meta, algoritmo, bEvalTime=True, h=None):
        pMisioneros = ProblemaMisioneros(estado_inicial, meta)
        if h: 
            if bEvalTime:
                print("Tiempo medio de ejecución: ")
                %timeit algoritmo(pMisioneros,h).solution()
                pMisioneros.analizados = 0
            sol = algoritmo(pMisioneros,h).solution()
        else: 
            if bEvalTime:
                print("Tiempo medio de ejecución: ")
                %timeit algoritmo(pMisioneros).solution()
                pMisioneros.analizados = 0
            sol = algoritmo(pMisioneros).solution()
        print("Solución: {0}".format(sol))
        print("Algoritmo: {0}".format(algoritmo.__name__))
        if h: 
            print("Heurística: {0}".format(h.__name__))
        else:
            pass
        print("Longitud de la solución: {0}. Nodos analizados: {1}".format(len(sol),pMisioneros.analizados))
        

In [5]:
# creamos un problema a partir de nuestra formalizacion de ProblemaMisioneros
# como parametros le pasamos el estado inicial, y el estado meta que esperamos
estado = ProblemaMisioneros((3, 3, 0), (0, 0, 1))

# le decimos a aima que resuelva nuestro problema con el metodo de busqueda en
# amplitud
#breadth_first_tree_search(estado).solution()
#%timeit breadth_first_tree_search(estado).solution()
ProblemaMisioneros.evalAlgorithm((3, 3, 0), (0, 0, 1), breadth_first_tree_search)
print("\n-----\n")
ProblemaMisioneros.evalAlgorithm((3, 3, 0), (0, 0, 1), breadth_first_graph_search)

Tiempo medio de ejecución: 
132 ms ± 12.5 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
Solución: [('2c', (0, 2)), ('1c', (0, 1)), ('2c', (0, 2)), ('1c', (0, 1)), ('2m', (2, 0)), ('1m1c', (1, 1)), ('2m', (2, 0)), ('1c', (0, 1)), ('2c', (0, 2)), ('1c', (0, 1)), ('2c', (0, 2))]
Algoritmo: breadth_first_tree_search
Longitud de la solución: 11. Nodos analizados: 11878

-----

Tiempo medio de ejecución: 
156 µs ± 11.8 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
Solución: [('2c', (0, 2)), ('1c', (0, 1)), ('2c', (0, 2)), ('1c', (0, 1)), ('2m', (2, 0)), ('1m1c', (1, 1)), ('2m', (2, 0)), ('1c', (0, 1)), ('2c', (0, 2)), ('1c', (0, 1)), ('2c', (0, 2))]
Algoritmo: breadth_first_graph_search
Longitud de la solución: 11. Nodos analizados: 15


In [6]:
#ProblemaMisioneros.evalAlgorithm((3, 3, 0), (0, 0, 1), depth_first_tree_search) INFINITO
ProblemaMisioneros.evalAlgorithm((3, 3, 0), (0, 0, 1), depth_first_graph_search)

Tiempo medio de ejecución: 
164 µs ± 29.6 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
Solución: [('1m1c', (1, 1)), ('1m', (1, 0)), ('2c', (0, 2)), ('1c', (0, 1)), ('2m', (2, 0)), ('1m1c', (1, 1)), ('2m', (2, 0)), ('1c', (0, 1)), ('2c', (0, 2)), ('1m', (1, 0)), ('1m1c', (1, 1))]
Algoritmo: depth_first_graph_search
Longitud de la solución: 11. Nodos analizados: 12


In [7]:
ProblemaMisioneros.evalAlgorithm((3, 3, 0), (0, 0, 1), uniform_cost_search)

Tiempo medio de ejecución: 
259 µs ± 50.9 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
Solución: [('1m1c', (1, 1)), ('1m', (1, 0)), ('2c', (0, 2)), ('1c', (0, 1)), ('2m', (2, 0)), ('1m1c', (1, 1)), ('2m', (2, 0)), ('1c', (0, 1)), ('2c', (0, 2)), ('1c', (0, 1)), ('2c', (0, 2))]
Algoritmo: uniform_cost_search
Longitud de la solución: 11. Nodos analizados: 15


b) Son las soluciones óptimas?
El algoritmo de coste uniforme y búsqueda en anchura tienen soluciones óptimas por definición (búsqueda en anchura si todos los operadores tienen el mismo coste) al recorrer el árbol por niveles no se alcanza nunca una solución que sea más profunda que otra.
El de profundidad encuentra también la solución óptima en este problema en específico, al ser la profundidad del árbol "pequeña" y estar la solución en las primeras ramas del arbol (por la izquierda).
La solución óptima es de profundidad 11 (11 acciones).

c) Coste de memoria de los algoritmos
En cuanto a búsqueda en anchura la diferencia está en que sin control de repetidos analiza 11878 nodos y su cola de abiertos es más grande, y en cambio con control de repetidos analiza 15 y los repetidos los guarda en su conjunto de cerrados. En medidas asintóticas tienen el mismo coste en memoria porque lo que te ahorras en la cola de abiertos se emplea en la cola de cerrados, pero en este problema en particular no (son muy pocos estados hasta la solución) y es mucho más beneficioso relizar control de repetidos.

Busqueda en profundidad ahorra al tener en memoria solo los nodos de una rama, que en este caso son muy pocos (12 nodos), y coste uniforme sigue el mismo principio que búsqueda en anchura con control de repetidos.

d) Que algoritmo es mejor
El mejor podría ser búsqueda en anchura con control de repetidos, en cuanto a su menor tiempo y espacio empleado.
Coste uniforme sería complicar la búsqueda en anchura al ser el coste de los caminos equivalente a la profundidad
en este problema, y búsqueda en profundidad con control de repetidos también funciona bien pero si el problema fuera 
más grande en profundidad del árbol de estados, como en el puzzle de ocho, daría solución suboptima en un tiempo mayor.

### Ejercicio opcional. Define alguna heurística y estudia las propiedades del algoritmo A*